### PR TITLE
Close #14313: Always add copy to clipboard action in share actions

### LINKED ISSUE #14313: 'Copy to Clipboard' option not present when 'Share' is clicked
## Steps to reproduce
Click on 'Share' option in the three dot menu when on a webpage

### Expected behavior
Apps that can share as an action appear along with 'Copy to Clipboard' https://bit.ly/3aZs1Sw

### Actual behavior
Recent apps appear, all actions appear, but 'Copy to Clipboard' does not. https://i.imgur.com/UO2PpyN.jpg

### Device & build information
* Device: Oneplus 7 / OxygenOS 10.3.4 / Android 10
* Build version: 80.0.1-beta.2 (Build #2015758611) and 79.0.5 (Build #2015758619)

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXIOS-818)

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt [hunk 1]
```diff
@@ -89,12 +94,32 @@ class ShareViewModel(application: Application) : AndroidViewModel(application) {
...
-            val shareAppsActivities = getIntentActivities(shareIntent, getApplication())
-            var apps = buildAppsList(shareAppsActivities, getApplication())
+            val shareAppsActivities = getIntentActivities(shareIntent, context)
+
+            var apps = buildAppsList(shareAppsActivities, context)
             recentAppsStorage.updateDatabaseWithNewApps(apps.map { app -> app.activityName })
             val recentApps = buildRecentAppsList(apps)
             apps = filterOutRecentApps(apps, recentApps)
 
+            val tint = ThemeManager.resolveAttribute(R.attr.secondaryText, context)
```
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:104]: just want to make sure this is the correct color. Do the icons use the same coloring as text in the share sheet?

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:104]: There's no other example of what colour to use here.  All other icons are retrieved from package manager (launcher icons)  I found that primary text colour was too bright and this is just right.

Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:104]: cc: @gabrielluong

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:104]: Let's send a screenshot of this over to Crystal and get a confirmation.

**Comment by MozillaNoah** [comment]
pending color confirmation, this looks good!
image

**Comment by jonalmeida** [comment]
> pending color confirmation, this looks good!

This screenshot is from the system share sheet. The share icon from the three-dot menu is our custom share sheet.

File: app/src/main/java/org/mozilla/fenix/share/ShareController.kt [hunk 2]
```diff
@@ -217,4 +227,16 @@ class DefaultShareController(
     private fun String.toDataUri(): String {
         return "data:,${Uri.encode(this)}"
     }
+
+    private fun copyClipboard() {
+        val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
+        val clipData = ClipData.newPlainText(getShareSubject(), getShareText())
+
+        clipboardManager.setPrimaryClip(clipData)
+        Toast.makeText(context, context.getString(R.string.toast_copy_link_to_clipboard), Toast.LENGTH_SHORT).show()
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/share/ShareController.kt:236]: I wasn't sure about this.  Shouldn't we try to emulate when we copy to clipboard?

File: app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt [hunk 3]
```diff
@@ -89,12 +94,32 @@ class ShareViewModel(application: Application) : AndroidViewModel(application) {
...
+                    AppShareOption(
+                        context.getString(R.string.share_copy_link_to_clipboard),
+                        copyIcon,
+                        ACTION_COPY_LINK_TO_CLIPBOARD,
+                        ""
+                    )
+
+                apps = listOf(copyApp) + apps
+            }
+
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:122]: Sure, I'll move it to a separate function.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -980,6 +980,10 @@
     <string name="sync_add_new_device_learn_button">Learn About Sending Tabs…</string>
     <!-- Text for the button to connect another device -->
     <string name="sync_add_new_device_connect_button">Connect Another Device…</string>
+    <!-- Text for copy link to the clipboard -->
+    <string name="share_copy_link_to_clipboard">Copy to clipboard</string>
+    <!-- Toast shown after copying link to clipboard -->
+    <string name="toast_copy_link_to_clipboard">Copied to clipboard</string>
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/res/values/strings.xml:986]: Sure, I'll move it up closer to the top of this section.

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -980,6 +980,10 @@
     <string name="sync_add_new_device_learn_button">Learn About Sending Tabs…</string>
     <!-- Text for the button to connect another device -->
     <string name="sync_add_new_device_connect_button">Connect Another Device…</string>
+    <!-- Text for copy link to the clipboard -->
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/res/values/strings.xml:983]: Sounds good, I'll update

**Comment by rocketsroger** [comment]
Demo

**Comment by rocketsroger** [comment]
Demo

File: app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt (hunk 1 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt]: Sure, please see attached screenshot.  I've also contacted Crystal.

**Comment by rocketsroger** [comment]
https://user-images.githubusercontent.com/6579541/157915444-498e64f1-aaca-400a-a8a5-867720e746c5.mp4

**Comment by rocketsroger** [comment]
https://user-images.githubusercontent.com/6579541/157915603-d7e291f1-57e7-48f8-93d0-0edb9fcb53ec.mp4

File: app/src/main/java/org/mozilla/fenix/share/ShareController.kt [hunk 6]
```diff
@@ -111,6 +120,7 @@ class DefaultShareController(
             when (e) {
                 is SecurityException, is ActivityNotFoundException -> {
                     snackbar.setText(context.getString(R.string.share_error_snackbar))
+                    snackbar.setLength(FenixSnackbar.LENGTH_LONG)
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/ShareController.kt:123]: Why make this one long?

File: app/src/main/java/org/mozilla/fenix/share/ShareController.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/share/ShareController.kt]: We spoke about this offline: TL;DR, if we're making the toast in our custom share sheet, we should follow our own design principles.

File: app/src/main/java/org/mozilla/fenix/share/ShareController.kt (hunk 6 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/share/ShareController.kt:123]: By default if we didn't specify length it was set to long.  So I kept this one as long.

**Comment by rocketsroger** [comment]
Latest update from Crystal's review.
demo_circular

File: app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt [hunk 7]
```diff
@@ -89,12 +92,18 @@ class ShareViewModel(application: Application) : AndroidViewModel(application) {
...
-            val shareAppsActivities = getIntentActivities(shareIntent, getApplication())
-            var apps = buildAppsList(shareAppsActivities, getApplication())
+            val shareAppsActivities = getIntentActivities(shareIntent, context)
+
+            var apps = buildAppsList(shareAppsActivities, context)
             recentAppsStorage.updateDatabaseWithNewApps(apps.map { app -> app.activityName })
             val recentApps = buildRecentAppsList(apps)
             apps = filterOutRecentApps(apps, recentApps)
 
+            /* if copy app is available, prepend to the list of actions */
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:102]:
```suggestion
            // if copy app is available, prepend to the list of actions
```

😛

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/share/ShareViewModel.kt:102]: Kidding of course, don't change this unless there is a real reason to edit this patch.

File: app/src/main/res/drawable/ic_share_clipboard.xml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<layer-list xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (jonalmeida) [inline app/src/main/res/drawable/ic_share_clipboard.xml:1]: Neat! This is a better solution!

**Review by jonalmeida — APPROVED** [review]
external-content duckduckgo com

**Pull request merged**